Remove Redux action types only referenced by action creators

aduth opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #102908

## Proposed Changes

* Updates lint script added in #102908 to further restrict action types to expect to be used by reducers or middlewares 


## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

As flagged in https://github.com/Automattic/wp-calypso/pull/102908#discussion_r2060779282, there appear to be a number of action types that are dispatched, but which have nothing listening to them, resulting in a lot of wasted effort resulting from dispatches, as well as inflating bundle size and cognitive load of our Redux application state.

The hope with the changes proposed here is that this lint script can assist in the elimination or migration of portions of the Redux state tree toward React Query alternatives.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Verify the lint script passes:

```
bin/find-unused-state-action-types.sh
```

Review changes for potential unintended impact.

For example, there were a few additional exemptions that needed to be made:

* The analytics middleware looked for `meta.analytics` on any dispatched action but did not look for specific action types. A couple of removal candidates were still relying on the analytics behavior.
* At least one action type was being referenced in `handlers.ts` files, which seem to be used in a way similar to middlewares ([example](https://github.com/Automattic/wp-calypso/blob/b9e88e93599d447d292e21f667bedb637dc88191/client/state/partner-portal/licenses/handlers.ts#L170-L178))

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

aduth committed:
Remove unused account recovery actions

aduth committed:
Remove unused current user fetch action

aduth committed:
Remove unused export action

aduth committed:
Remove unused invite action

aduth committed:
Permit actions referenced in reducer naming variation

aduth committed:
Permit actions referenced in "handlers"

aduth committed:
Remove unused JITM_OPEN_HELP_CENTER

Added in #70884, but unclear that it's ever relevant to opening help center

aduth committed:
Remove unused Keyring action types

aduth committed:
Remove unused Lasagna action type

aduth committed:
Remove unused Mailchimp action type

aduth committed:
Remove unused memberships action types

aduth committed:
Remove unused post action types

aduth committed:
Remove unused push notifications action type

aduth committed:
Remove unused site keyrings action types

aduth committed:
Remove unused email stats action types

aduth committed:
Add explicit references to action types with meta analytics

aduth commented:
I've rebased this against `trunk` after the merge of #102908 and marking it ready for review. This touches a lot of code, and I think the biggest thing here is checking my assumptions that an action type (and any action creators which dispatch it) are effectively useless unless there's some reducer or middleware that's paying attention to it.

aduth commented on client/state/analytics/middleware.js:
```diff
@@ -68,6 +70,11 @@ export const analyticsMiddleware = () => ( next ) => ( action ) => {
+			dispatcher( action );
+			break;
+
+		case GRAVATAR_RECEIVE_IMAGE_FAILURE:
```
Including specific action types here defeats the purpose of the middleware being agnostic to specific action types and listening for any action with `meta.analytics`. However, I personally feel that's a pattern we should move away from toward explicit analytics behavior, since the previous approach makes it much harder to track which action types trigger analytics. This is evidenced by the extra work it caused me in this pull request, and the riskiness surrounding these changes.

Ideally, these changes would only be temporary until we find a way to move the analytics behavior to be more explicit with the action types currently using them.

As an alternative, I'd be open to reverting the changes to this file and compiling an allowlist of certain action types that look unused but actually are.

tyxla replied:
Yep, let's address that in a separate PR.

jsnajdr replied:
Maybe I'm missing something, but seems to me we don't need neither of these two new `case`s.

The `GRAVATAR_RECEIVE_IMAGE_FAILURE` action is a complicated object created with `withAnalytics` and `composeAnalytics`. In `receiveGravatarImageFailed`, we can replace this with three plain dispatches:
```js
dispatch( recordTracksEvent( 'calypso_edit_gravatar_file_receive_failure' ) );
dispatch( bumpStat( 'calypso_gravatar_update_error', statName ) );
dispatch( errorNotice( errorMessage, { id: 'gravatar-upload' } ) );
```
That's all we need, and it leaves `GRAVATAR_RECEIVE_IMAGE_FAILURE` completely unused. It looks easy enough to be directly in this PR.

And the `ANALYTICS_MULTI_TRACK` action? It seems completely covered by the default `if ( action.meta?.analytics )` condition. Every `ANALYTICS_MULTI_TRACK` action ever created has the `meta.analytics` field, that's how `composeAnalytics` works.

aduth replied:
Is the idea with `composeAnalytics` to reduce the number of `dispatch` calls? I don't see much benefit outside of that, and it's surely a lot simpler code to just dispatch them directly, like you suggested.

It leaves me wondering how we'd ideally want this all to work, since `composeAnalytics` is widely used and I can imagine a developer in the future encountering another case like this one, and being confused how to address.

I'd wonder if we should take it to one of the two extremes:

- Remove `composeAnalytics` altogether and always dispatch individual events
- List all of the action types that include composed analytics metadata in this file as a canonical listing, so we don't encounter a scenario where an action type might otherwise appear "unused"

jsnajdr replied:
Yes, the idea of `composeAnalytics` and `withAnalytics` is to take an existing action and add analytics as an "attachment", i.e., the `meta.analytics` property. In the Redux store, the analytics middleware will take care of `meta.analytics` and then will send the action down the middleware chain, all the way to the reducer. It will be all processed as one dispatch.

If you use `composeAnalytics` without `withAnalytics`, you're not attaching analytics to an existing action but you're creating an analytics-only action. Typicall to record two or more analytics events. This one then has the "placeholder" type `ANALYTICS_MULTI_TRACK`. This action type doesn't need to be detected anywhere: the analytics middleware will process the `meta.analytics` field, and then the action will be processed by the rest of the store without any reaction.

Of course, all this can be achieved by simply dispatching all the actions individually, or packing them into a thunk:
```js
dispatch( ( d ) => {
  d( performSomething( siteId ) );
  d( recordGoogleEvent( 'foo' ) );
  d( recordTracksEvent( 'bar' ) );
} );
```
The only downside is that all three actions need to go through the Redux dispatch. Running all the reducers, only to discover that there is not state change. The composed analytics do this only once.

But this performance benefit is easy to recover: in the analytics middleware, intercept the `ANALYTICS_MULTI_TRACK` action and stop the processing there: don't call `next()`. Then it won't waste time going through the other middlewares and the reducer.

What's the future of `composeAnalytics` and `withAnalytics`? I'd prefer to remove them completely in favor of dispatching the actions individually. There is a lot of usages (around 100) but with some AI assistance and/or AI-generated codemod, the migration shouldn't be hard. Transform the usages either to a series of `dispatch` calls, or into a thunk with a series of `dispatch` calls.

aduth replied:
> What's the future of `composeAnalytics` and `withAnalytics`? I'd prefer to remove them completely in favor of dispatching the actions individually. There is a lot of usages (around 100) but with some AI assistance and/or AI-generated codemod, the migration shouldn't be hard. Transform the usages either to a series of `dispatch` calls, or into a thunk with a series of `dispatch` calls.

I'd be in favor of that, and it'd make the code a lot simpler to follow 👍 

Replaced the one instance here with individual dispatches in 8bc3c586bc4.

tyxla commented on client/blocks/jitm/index.jsx:
```diff
@@ -106,15 +105,7 @@ function getEventHandlers( props, dispatch ) {
 			dispatch( recordTracksEvent( tracks.click.name, { ...tracks.click.props, ...eventProps } ) );
 		}
 		if ( jitm.action ) {
-			switch ( jitm.action.type ) {
-				// Cases for dispatching action thunks
-				case JITM_OPEN_HELP_CENTER:
-					dispatch( openHelpCenterFromJITM( jitm.action.payload ) );
-					break;
-				default:
-					// Dispatch regular actions
-					dispatch( jitm.action );
-			}
```
Nice find 👍

tyxla commented on client/sites/marketing/connections/service.jsx:
```diff
@@ -450,27 +447,9 @@ export class SharingService extends Component {
 			// At this point, if there are no available accounts to
 			// select, we must assume the user closed the popup
 			// before completing the authorization step.
-			this.props.failCreateConnection( {
-				message: this.props.translate(
-					'The %(service)s connection could not be made because no account was selected.',
-					{
-						args: { service: this.props.service.label },
-						context: 'Sharing: Publicize connection confirmation',
-					}
-				),
-			} );
```
Have we tested these? Is it possible we're not dispatching notices in these cases? This looks like something that belonged to the notices middleware, but we refactored at some point (probably in PRs like https://github.com/Automattic/wp-calypso/pull/49261 and eventually https://github.com/Automattic/wp-calypso/pull/50035 which removed the entire middleware)

aduth replied:
> Have we tested these? Is it possible we're not dispatching notices in these cases? This looks like something that belonged to the notices middleware, but we refactored at some point (probably in PRs like #49261 and eventually #50035 which removed the entire middleware)

Thus far I've only made changes based on references in code, but I'll try to run through this flow next week to see how it behaves.

tyxla replied:
Sounds good, if we can confirm that we aren't missing to display some notices to the user, I think this PR should be good to go 👍

jsnajdr replied:
The `failCreateConnection` actions are broken and should be fixed instead of removing them. The errors are easy to simulate:
- add a logpoint to the `failCreateConnection` action creator and log when it's called
- go to `marketing/connections/:site` and try to connect to Facebook or something else
- reject/close the popup that opens, that makes the connection fail
- the logpoint is triggered: there is an error message when the Facebook connection failed
- but the UI doesn't say anything!

Here's the error I see:
<img width="1167" alt="Screenshot 2025-05-12 at 12 37 15" src="https://github.com/user-attachments/assets/b863e267-814f-44ca-bf7d-674bd80955b9" />

We should refactor the noop `failCreateConnection` dispatches to `createNotice`.

> This looks like something that belonged to the notices middleware, but we refactored at some point

I think it's not because of the notices middleware: the `failCreateConnection` action probably never worked. Look at #11268, a very old PR by @obenland. That's where `failCreateConnection` was introduced and it says "prepare Service module for Redux switch". It only _prepared_ the error handling but then it was never implemented. Let's finish it now after 8 years, like a gothic cathedral 🙂

tyxla replied:
Thanks for the sleuthing, @jsnajdr. Let's use the opportunity to fix these then, rather than deleting them 😉

aduth replied:
Great find @jsnajdr . Glad that our cleanup could also highlight a real user bug ❤️ 

I'll create a separate pull request to isolate a fix on the missing error notice.

aduth replied:
Fix PR: #103360

tyxla commented on client/state/memberships/gifts/actions.js:
```diff
@@ -68,12 +43,6 @@ export const requestAddGift = ( siteId, gift, noticeText, onConfirm ) => {
 
 export const requestDeleteGift = ( siteId, giftId, noticeText ) => {
```
Noting that some of this code is actually completely removable as it's not used, not just the non-referenced actions. Try looking for `requestDeleteGift` for example.

aduth replied:
> Noting that some of this code is actually completely removable as it's not used, not just the non-referenced actions. Try looking for `requestDeleteGift` for example.

Nice find, even more to remove! I double-checked all the updated action creators and I think it's just this one that's unreferenced currently. Updated in 97bc95f2199.

tyxla commented on client/state/posts/actions/save-post.js:
```diff
@@ -34,14 +34,7 @@ export function savePost( siteId, postId = null, post, silent = false ) {
-					error,
-				} );
-			}
+			() => {}
```
Should we just remove this one?

aduth replied:
Omitting is not quite the same, because then it becomes an unhandled exception.

Maybe that's how it _should_ be, but this keeps consistency with how it effectively worked previously.

tyxla replied:
Got it 👍

tyxla commented on client/state/posts/actions/trash-post.js:
```diff
@@ -40,14 +40,7 @@ export function trashPost( siteId, postId, silent = false ) {
-					error,
-				} );
-			}
+			() => {}
```
Should we just omit?

tyxla reviewed (commented):
This is looking great, thanks for the great cleanup! 

Just a few minor comments, let me know what you think. 

Will need a rebase before 🚢

aduth committed:
Remove unreferenced action creator

tyxla reviewed (approved):
Nice. Couldn't spot any further problems.

LGTM 🚀

jsnajdr commented on client/blocks/jitm/index.jsx:
```diff
@@ -106,15 +105,7 @@ function getEventHandlers( props, dispatch ) {
-					// Dispatch regular actions
-					dispatch( jitm.action );
-			}
+			dispatch( jitm.action );
```
This change looks wrong. This dispatch will dispatch action object with type `JITM_OPEN_HELP_CENTER` and the action will do absolutely nothing, because the constant is no longer even defined.

We need to keep calling the `openHelpCenterFromJITM` thunk/function, which dispatches actions in the `help-center` `@wordpress/data` store.

The `openHelpCenterFromJITM` doesn't need to be a thunk any more, it can be a regular function, because the `JITM_OPEN_HELP_CENTER` Redux dispatch can be removed from it.

aduth replied:
@jsnajdr I'm not sure I follow how the `JITM_OPEN_HELP_CENTER` case statement would ever be reached?

Are there steps to reproduce how it can be reached in trunk?

jsnajdr replied:
It could be reached if the `jitm` object (returned by a REST endpoint) had an `action` field with `type: 'JITM_OPEN_HELP_CENTER'`.

But it turns out that never happens! The JITM that returned the was removed in D101077-code. And that was the only usage of the `with_redux_action` option for constructing JITMs, so this entire `if ( jitm.action )` check can be removed.

We should also remove the unused `with_redux_action` code from the backend. IMO it was never a good idea to couple the backend so closely with Calypso implementation details and a specific Redux action.

aduth replied:
> But it turns out that never happens! The JITM that returned the was removed in D101077-code. And that was the only usage of the `with_redux_action` option for constructing JITMs, so this entire `if ( jitm.action )` check can be removed.

Good find! I wasn't aware we had actions being handled from REST API responses like this.

Removed the entire `if` condition in a3ebbb2bae7.

jsnajdr commented on client/state/account-recovery/settings/test/actions.js:
```diff
@@ -357,40 +330,6 @@ describe( 'account-recovery actions', () => {
-					error: expect.objectContaining( errorResponse ),
-				} )
-			),
-	} );
```
Ideally we should modify the `postCondition`s instead of deleting them. The tested thunks dispatch the `NOTICE_CREATE` actions on success/failure, with a certain `status` and `text` that can be tested.

aduth replied:
> Ideally we should modify the `postCondition`s instead of deleting them. The tested thunks dispatch the `NOTICE_CREATE` actions on success/failure, with a certain `status` and `text` that can be tested.

Restored tests for notices in d008693ab0c + d89b31c1fe1

jsnajdr commented on client/state/action-types.ts:
```diff
@@ -902,14 +873,12 @@ export const SITE_IMPORTER_IS_SITE_IMPORTABLE_START = 'SITE_IMPORTER_IS_SITE_IMP
 export const SITE_KEYRINGS_REQUEST_FAILURE = 'SITE_KEYRINGS_REQUEST_FAILURE';
 export const SITE_KEYRINGS_REQUEST_SUCCESS = 'SITE_KEYRINGS_REQUEST_SUCCESS';
 export const SITE_KEYRINGS_SAVE_SUCCESS = 'SITE_KEYRINGS_SAVE_SUCCESS';
 export const SITE_KEYRINGS_UPDATE = 'SITE_KEYRINGS_UPDATE';
```
The `SITE_KEYRINGS_UPDATE` action is also unused. We ever react only to `_UPDATE_SUCCESS`.

The same applies to `SITE_KEYRINGS_DELETE`

aduth replied:
> The `SITE_KEYRINGS_UPDATE` action is also unused. We ever react only to `_UPDATE_SUCCESS`.
> 
> The same applies to `SITE_KEYRINGS_DELETE`

I'll check why those aren't being flagged by the current script.

aduth replied:
Ah, I think this is because we're naively checking for an occurrence of the string, and in this case there's a suffixed version that's being used. This is another case where it'd be nice if we could actually parse and look at the actual tokens being used.

Alternatively, maybe we could update the pattern to look for some sort of break character like `,` or ` ` (space) following the action name, assuming it'd always appear like one of the following:

```ts
import { SITE_KEYRINGS_UPDATE } from 'calypso/state/action-types';
import { SITE_KEYRINGS_UPDATE, SOMETHING_ELSE } from 'calypso/state/action-types';
import {
	SOMETHING_ELSE,
	SITE_KEYRINGS_UPDATE,
} from 'calypso/state/action-types';
```

aduth replied:
I improved the script in 14a5a5835ef and b53e0c9b96f (the latter likely a rebase mishap, since it was the original purpose of the pull request). Now it's catching a lot more unused action types, including `SITE_KEYRINGS_UPDATE` and `SITE_KEYRINGS_DELETE`. The build should be failing now until I resolve the remaining references.

jsnajdr commented on client/state/data-layer/wpcom/me/notification/settings/index.js:
```diff
@@ -40,7 +31,6 @@ registerHandlers( 'state/data-layer/wpcom/me/notification/settings/index.js', {
 	[ NOTIFICATION_SETTINGS_REQUEST ]: [
 		dispatchRequest( {
 			fetch: requestNotificationSettings,
-			onSuccess: updateSettings,
```
This is strange, why would we fetch notification settings if we ignore the response?

It seems that this entire `NOTIFICATION_SETTINGS_REQUEST` data layer handler and the associated `requestNotificationSettings` action creator is never used except in tests. Can be deleted.

Instead we use the `fetchSettings` thunk with the `NOTIFICATION_SETTINGS_FETCH`. It fetches from exactly the same `/me/notifications/settings` endpoint. And is used.

aduth replied:
Yeah I think there's a lot to clean up here with the notifications behavior. I might split that out to a separate pull request to keep this one reasonably-sized.

**Edit:** See #103608

aduth committed:
Enforce action types not referenced in actions alone

aduth committed:
Enforce break character to exclude suffixed references

aduth committed:
Expand script description about assumptions made

aduth committed:
Restore tests for expected notices on account recovery results

jsnajdr commented on client/state/mailchimp/settings/actions.js:
```diff
@@ -4,7 +4,6 @@ import {
 	MAILCHIMP_SETTINGS_RECEIVE,
 	MAILCHIMP_SETTINGS_UPDATE,
 	MAILCHIMP_SETTINGS_UPDATE_SUCCESS,
-	MAILCHIMP_SETTINGS_UPDATE_FAILURE,
```
Two opportunities for more cleanup:
- the `receiveSettings` action creator in this module is not used anywhere. Either remove it, or use it in the dispatches that construct a `MAILCHIMP_SETTINGS_RECEIVE` action explicitly, without a creator.
- the `MAILCHIMP_SETTINGS_UPDATE_SUCCESS` is identical to `MAILCHIMP_SETTINGS_RECEIVE`, the reducer handles them the same way. That's an opportunity to remove `MAILCHIMP_SETTINGS_UPDATE_SUCCESS`.

aduth replied:
> * the `receiveSettings` action creator in this module is not used anywhere. Either remove it, or use it in the dispatches that construct a `MAILCHIMP_SETTINGS_RECEIVE` action explicitly, without a creator.

This is making me think it'd be valuable to also have checks for use of exported action creators elsewhere in the application. I've now seen several examples of action creators that exist but are never used.

aduth replied:
> * the `receiveSettings` action creator in this module is not used anywhere. Either remove it, or use it in the dispatches that construct a `MAILCHIMP_SETTINGS_RECEIVE` action explicitly, without a creator.

Removed in 04fbfeab957.

> * the `MAILCHIMP_SETTINGS_UPDATE_SUCCESS` is identical to `MAILCHIMP_SETTINGS_RECEIVE`, the reducer handles them the same way. That's an opportunity to remove `MAILCHIMP_SETTINGS_UPDATE_SUCCESS`.

Consolidated in fe21abe1347.

aduth committed:
Add middleware-like data-layer as exempted path

aduth committed:
Remove unused Mailchimp action type

aduth committed:
Remove unused Publicize action types

aduth committed:
Correctly define and use stored cards as constants

aduth committed:
Remove unused keyring action type

aduth committed:
Remove unused receiveSettings action creator

aduth committed:
Consolidate MAILCHIMP_SETTINGS_UPDATE_SUCCESS with MAILCHIMP_SETTINGS_RECEIVE

aduth committed:
Replace composed Gravatar failure analytics with individual dispatches

See: https://github.com/Automattic/wp-calypso/pull/103138#discussion_r2099963202

aduth committed:
Remove unreachable JITM action handling

The JITM that returned the was removed in D101077-code. And that was the only usage of the with_redux_action option for constructing JITMs, so this entire if ( jitm.action ) check can be removed.

See: https://github.com/Automattic/wp-calypso/pull/103138#discussion_r2104444882

aduth committed:
Remove unused Keyring action type

jsnajdr commented on client/state/partner-portal/stored-cards/actions.ts:
```diff
@@ -23,22 +33,22 @@ export const fetchStoredCards =
 				dispatch( {
-					type: 'STORED_CARDS_UPDATE_IS_PRIMARY_COMPLETED',
+					type: STORED_CARDS_UPDATE_IS_PRIMARY_COMPLETED,
 					payment_method_id: data.items.find( ( currCard ) => currCard.is_default ),
```
Here the `payment_method_id` should be the `.id` field of the found item, not the entire item (of type `PaymentMethod`.

The `HAS_MORE_ITEMS`, `ITEMS_PER_PAGE` and `FETCH_COMPLETED` actions could be merged into one. The current setup is very unusual: piece of `data` are dispatched as separate actions, and for each action there is a single sub-reducer that reacts to it.

Not a blocker for this PR, but wanted to point out what I noticed.

jsnajdr commented on client/state/partner-portal/stored-cards/reducer.ts:
```diff
@@ -36,73 +47,77 @@ export const items: Reducer< PaymentMethod[], ItemsAction > = ( state = [], acti
 	perPage: number;
 };
 
 export const itemsPerPage: Reducer< number, ItemsPerPageAction > = ( state = 30, action ) => {
```
The types of these sub-reducers are wrong. TypeScript is made to think that `ItemsPerPageAction` is the only action type the reducer is ever called with, and that the code in the `default` case is unreachable.

aduth replied:
> The types of these sub-reducers are wrong. TypeScript is made to think that `ItemsPerPageAction` is the only action type the reducer is ever called with, and that the code in the `default` case is unreachable.

I spent quite a bit of time trying to figure out what the "correct" way to type a reducer like this is, since I suspect the reason these actions were referenced as strings in the first place was to work around how TypeScript will detect code as unreachable except in the `default:` case (even though the `default:` is equally "unreachable" in the same way). Looking around our code, I see a lot of different ways that we type reducers, and most all of them seem to be workarounds or redundant conditions (`if` + `if` + `if`, rather than `switch` or `if` + `else if`).

jsnajdr reviewed (approved):
I finally made it all the way to the end of the diff 😆 Looks shippable now.
